docs/HowToFAQ/Errors-Refinement.md

```diff
@@ -80,78 +80,68 @@ module Q refines P {
 
 A submodule M within a module that is refining some base module must refine some submodule M in the base module.
 
-## **Error: type declaration '_name_' is not allowed to change the requirement of supporting equality** {#ref_mismatched_equality}
-
-<!-- TODO -->
-
-## **Error: type declaration '_name_' is not allowed to change the requirement of supporting auto-initialization** {#ref_mismatched_auto_init}
-
-<!-- TODO -->
-
-## **Error: type declaration '_name_' is not allowed to change the requirement of being nonempty** {#ref_mismatched_nonempty}
-
-<!-- TODO -->
-
-## **Error: a type declaration that requires equality support cannot be replaced by this trait** {#ref_trait_mismatched_equality}
-
-<!-- TODO -->
-
-## **Error: a type declaration that requires equality support cannot be replaced by a codatatype** {#ref_equality_support_precludes_codatatype}
+## **Error: to be a refinement of _kind_ '_name_' declared with (==), _kind_ '_name_' must support equality {#ref_mismatched_type_characteristics}
 
 ```dafny
 module P {
   type T(==)
 }
 module Q refines P {
-  codatatype T = A
+  type T = AlwaysAndForeverMore
+  codatatype AlwaysAndForeverMore = Cons(int, AlwaysAndForeverMore)
 }
 ```
 
-Codatatypes do not generally support equality and so cannot be refinements of an abstract type that declares that it supports equality.
+The abstract type `T` in module `P` says it supports equality, but its attempted refinement in module `Q` does not.
+Codatatypes do not generally support equality and so cannot be refinements of equality-supporting abstract types.
 
-## **Error: type '_name_', which does not support equality, is used to refine an abstract type with equality support** {#ref_mismatched_type_equality_support}
+## **Error: to be a refinement of _kind_ '_name_' declared with (!new), _kind_ '_name_' must contain no references {#ref_mismatched_type_characteristics}
```

Comment:
IDs declared in curly braces must be unique. Could you either regroup these sections together (so that the user knows there is no other section with the same ID), or make separate error IDs in the code? I'm fine with both, although I'd slightly prefer the second since you worked hard on making each example unique.